internal/destinations/s3/s3.go

```diff
@@ -0,0 +1,144 @@
+package runreveal
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/runreveal/kawa"
+	"github.com/runreveal/kawa/internal/types"
+	batch "github.com/runreveal/kawa/x/batcher"
+)
+
+type Option func(*s3)
+
+func WithBucketName(bucketName string) Option {
+	return func(s *s3) {
+		s.bucketName = bucketName
+	}
+}
+
+func WithPathPrefix(pathPrefix string) Option {
+	return func(s *s3) {
+		s.pathPrefix = pathPrefix
+	}
+}
+
+func WithBucketRegion(bucketRegion string) Option {
+	return func(s *s3) {
+		s.bucketRegion = bucketRegion
+	}
+}
+
+func WithCustomEndpoint(customEndpoint string) Option {
+	return func(s *s3) {
+		s.customEndpoint = customEndpoint
+	}
+}
+
+func WithAccessKeyID(accessKeyID string) Option {
+	return func(s *s3) {
+		s.accessKeyID = accessKeyID
+	}
+}
+
+func WithAccessSecretKey(accessSecretKey string) Option {
+	return func(s *s3) {
+		s.accessSecretKey = accessSecretKey
+	}
+}
+
+type s3 struct {
+	batcher *batch.Destination[types.Event]
+
+	bucketName   string
+	bucketRegion string
+	pathPrefix   string
+
+	customEndpoint  string
+	accessKeyID     string
+	accessSecretKey string
+}
+
+func New(opts ...Option) *s3 {
+	ret := &s3{}
+	for _, o := range opts {
+		o(ret)
+	}
+	ret.batcher = batch.NewDestination[types.Event](ret,
+		batch.FlushLength(25),
```

Comment:
25 is probably a bit too low for a sane default, and let's expose this through options as well.